modules/renter/downloadcache.go

```diff
@@ -4,72 +4,152 @@ package renter
+
+// Swap swaps two elements from the heap
+func (cch chunkCacheHeap) Swap(i, j int) {
+	cch[i], cch[j] = cch[j], cch[i]
+	cch[i].index = i
+	cch[j].index = j
+}
+
+// Push adds an element to the heap
+func (cch *chunkCacheHeap) Push(x interface{}) {
+	n := len(*cch)
+	chunkData := x.(*chunkData)
+	chunkData.index = n
+	*cch = append(*cch, chunkData)
+}
+
+// Pop removes element from the heap
+func (cch *chunkCacheHeap) Pop() interface{} {
+	old := *cch
+	n := len(old)
+	chunkData := old[n-1]
+	chunkData.index = -1 // for safety
+	*cch = old[0 : n-1]
+	return chunkData
+}
+
+// update updates the heap and reorders
+func (cch *chunkCacheHeap) update(cd *chunkData, id string, data []byte, lastAccess time.Time) {
+	cd.id = id
+	cd.data = data
+	cd.lastAccess = lastAccess
+	heap.Fix(cch, cd.index)
+}
+
+// add adds the chunk to the cache if the download is a streaming
 // endpoint download.
 // TODO this won't be necessary anymore once we have partial downloads.
-func (udc *unfinishedDownloadChunk) addChunkToCache(data []byte) {
-	if udc.download.staticDestinationType != destinationTypeSeekStream {
+func (dcc *downloadChunkCache) add(data []byte, cacheID string, destinationType string) {
+	if destinationType != destinationTypeSeekStream {
 		// We only cache streaming chunks since browsers and media players tend
 		// to only request a few kib at once when streaming data. That way we can
 		// prevent scheduling the same chunk for download over and over.
 		return
 	}
-	udc.cacheMu.Lock()
-	defer udc.cacheMu.Unlock()
+	dcc.mu.Lock()
+	defer dcc.mu.Unlock()
 
 	// Prune cache if necessary.
-	for len(udc.chunkCache) >= downloadCacheSize {
-		var oldestKey string
-		oldestTime := time.Now()
-
-		// TODO: turn this from a structure where you loop over every element
-		// (O(n) per access) to a min heap (O(log n) per access).
-		// currently not a issue due to cache size remaining small (<20)
-		for id, chunk := range udc.chunkCache {
-			if chunk.lastAccess.Before(oldestTime) {
-				oldestTime = chunk.lastAccess
-				oldestKey = id
-			}
+	if dcc.cacheSize == 0 {
+		build.Critical("DownloadCacheSize is set to zero")
+		return
+	}
+
+	for len(dcc.chunkCacheMap) >= int(dcc.cacheSize) {
+		// Remove from Heap
+		cd := heap.Pop(&dcc.chunkCacheHeap).(*chunkData)
+
+		// Remove from Map
+		if _, ok := dcc.chunkCacheMap[cd.id]; !ok {
+			build.Critical("Cache Data chunk not found in chunkCacheMap.")
 		}
-		delete(udc.chunkCache, oldestKey)
+		delete(dcc.chunkCacheMap, cd.id)
 	}
 
-	udc.chunkCache[udc.staticCacheID] = &cacheData{
+	// Add chunk to Map and Heap
+	cd := &chunkData{
+		id:         cacheID,
 		data:       data,
 		lastAccess: time.Now(),
 	}
+	dcc.chunkCacheMap[cacheID] = cd
+	heap.Push(&dcc.chunkCacheHeap, cd)
+	dcc.chunkCacheHeap.update(cd, cd.id, cd.data, cd.lastAccess)
+}
+
+// init initializes the downloadChunkCache
+func (dcc *downloadChunkCache) init() {
```

Comment:
Just for the sake of completeness `init` should be `Init`, be threadsafe (acquire lock) and it should probably contain a sanity check (`build.Critical`) to make sure we can't call init twice on the same cache.
`heap.Init` is idempotent but our custom `Init` isn't.